Bug 1338941 - web page for new update out of date dashboard

RobertTStrong opened the pull request:
This is in conjunction with the following changes to mozilla-services/data-pipeline
https://github.com/mozilla-services/data-pipeline/pull/242

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/telemetry-dashboard/286)
<!-- Reviewable:end -->

Commit:
Bug 1338941 - web page for new update out of date dashboard

chutten commented on update-orphaning/index-old.html:
```diff
@@ -0,0 +1,513 @@
+<!DOCTYPE html>
```
There doesn't appear to be anything that was in the old dash that isn't in the new one, so just remove the old one instead of keeping it around.

RobertTStrong replied:
Before deciding to copy it I discussed this with ddurst and decided that it should be kept for awhile since it has data going back further than this dashboard. We'll evaluate removing it sometime in the future.

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
 <html>
 <head>
-<title>Update orphaning dashboard</title>
+<title>Application Update Out Of Date Dashboard</title>
```
Prepend a "Firefox" on this?

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
 <script type="text/javascript" src="https://telemetry.mozilla.org/new-pipeline/lib/d3pie.min.js"></script>
 <style>
-
 body {
```
We can take the opportunity to pull the CSS out into its own file.

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+.p0_title,
+.p1_title,
+.p2_title,
+.p3_title {
```
The textbook way to do this would be to make a `.bold-title` class and then apply that class to each of these elements.

RobertTStrong replied:
These are d3pie generated elements that don't have a method to set the class when they are generated.

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
 body {
   font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
+}
+#mainDiv {
```
Textbook naming conventions for CSS classes and ids are `hyphenated-lower`

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+<select id="weeklyDropdown" onchange="populateDashboard()"><option>--/--/----</option></select>
+</div>
+<div id="mainDiv">
+<h1>Application Update Out Of Date Dashboard</h1>
```
s/Application/Firefox/ ?

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+The Spark Notebook for the weekly reports is scheduled to run every Sunday and can be found <a href="https://github.com/mozilla-services/data-pipeline/tree/master/reports/update-orphaning/Update orphaning analysis using longitudinal dataset.ipynb">here</a>.<br/>
+This web page can be found <a href="https://github.com/mozilla/telemetry-dashboard/blob/gh-pages/update-orphaning/index.html">here</a>.
+</div>
 <script>
```
We can take the opportunity to pull out the script to its own file

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+  var bcTransDur = 800;
+  var bcMargin = {top: 20, right: 20, bottom: 30, left: 40};
+  var bottomOffset = 60;
+  var bcWidth = 960 - bcMargin.left - bcMargin.right;
```
Ugh, I hate magic numbers. SVG never seemed particularly fun to work with.

RobertTStrong replied:
Added comments.

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
-    daysToLastWed = 11;
-  } else {
-    daysToLastWed = 4 + todaysWeekday;
+  var bcTransDur = 800;
```
JS constants should be `CAPS_WITH_UNDERSCORES`

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+  function getSortedData(aData, aSectionPrefix) {
+    return aData.sort(document.getElementById(aSectionPrefix + "SortInput").checked
+               ? function(a, b) { var val = b.value - a.value; if (val) { return val; } return d3.ascending(a.index, b.index); }
+               : function(a, b) { return d3.ascending(a.index, b.index); })
```
For future reference, in the cases when you aren't using `this`, there are lambda constructs that can save some typing:
```javascript
(a, b) => d3.ascending(a.index, b.index)
```

RobertTStrong replied:
Skipped in favor of consistency with the surrounding code.

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+    transition.selectAll("rect." + aSectionPrefix)
+      .attr("x", function(d) { return x(d.index); });
+
+    transition.select("g.x.axis." + aSectionPrefix + "Axis")
```
Seems to be a duplicate of line 283. Should probably be pulled out into its own function

RobertTStrong replied:
Fixed

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
-  friAfterLastWedDate.setDate(lastWedDate.getDate() + 2);
-  var lastWedDataFile = [year, month, date, ".json"].join("");
+
+  const checkCodeGeneralDetails = {
```
Might be nice to throw some block comments atop these consts to tell future devs where they came from, like how these come from UpdateTelemetry.jsm, and the next bunch come from... actually I'm not sure where those come from.

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+    "100": {code: "Invalid code", desc: "Unknown"}};
+
+  const checkCodeGood = [1, 2, 3, 4, 6, 7, 8, 9, 10, 11];
+  const checkCodeBad = [0, 5, 14, 15, 18, 19, 20, 21, 22, 23, 24, 25, 27, 29, 30, 31, 32, 33, 34, 35, 36];
```
And as before and throughout, consts should have names that adhere to style.

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+    .scale(checkCodeX)
+    .orient("bottom");
+
+  var checkCodeYAxis = d3.svg.axis()
```
If I'm not mistaken, these can all be const

RobertTStrong replied:
Done and I was able to clean this up significantly

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+
+  stateFailureCodeStartupSVG.append("g")
+    .attr("class", "y axis stateFailureCodeStartupAxis");
+
```
Is it worth trying to figure out a way to parameterize these bricks of similar-yet-distinct code? It seems as though a change in one would need to be repeated throughout, but they're _just_ different enough I'm not sure if it's worth it.

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
-  startDate.setDate(2);
-  startDate.setMonth(2);
+  startDate.setDate(23);
+  startDate.setMonth(9);
```
I'd love a comment about what makes Sept 23, 2016 special.

chutten replied:
Or is it October?

RobertTStrong replied:
Added comments

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+    if (day < 10) {
+      day = "0" + day;
+    }
+    return [year, month, day];
```
The types of month and day will be numbers if > 9, strings if not.

RobertTStrong replied:
Made the function always return a string

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+        break;
+      }
+    }
+    var total = aJSONData[aIndex] ? aJSONData[aIndex] : 0;
```
commonly written as `aJSONData[aIndex] || 0`

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+      detailsText = detailsText + " Unknown</b>: Unknown";
+    }
+    document.getElementById(aPrefixID + "Link").innerHTML = linkText;
+    document.getElementById(aPrefixID + "Details").innerHTML = detailsText;
```
Ooooh, I'm not very happy to see complicated strings of questionable text being added using innerHTML. This is how you get script injection.

I would prefer it if you created the elements with DOM methods like `document.createElement('b')` and then inserted the text in its inert form using `Element.textContent = text`

RobertTStrong replied:
Done

chutten commented on update-orphaning/index.html:
```diff
@@ -1,206 +1,1086 @@
+      try {
+        result = aData[aName][aKey];
+      } catch (e) {
+      }
```
Should at least console.warn, no?

RobertTStrong replied:
Added. Note: this is so it is possible to add new data without breaking reports using old data

chutten commented on update-orphaning/index.html:
```diff
@@ -261,253 +1150,299 @@
+      });
+      updateBarChart(stateFailureCodeStartupData, "stateFailureCodeStartup", stateFailureCodeStartupBarClassFn, stateFailureCodeStartupX,
+                     stateFailureCodeStartupY, stateFailureCodeStartupXAxis, stateFailureCodeStartupYAxis,
+                     stateFailureCodeStartupOfConcern, stateFailureCodeGeneralDetails, null, false);
```
If we can somehow parameterize above, this will be cake.

RobertTStrong replied:
Cleaned up.

chutten reviewed (commented):
I adore the new look and the additional levels of detail.

Commit:
fix review comments for Bug 1338941 - web page for new update out of date dashboard

Commit:
fix review comments part 2 for Bug 1338941 - web page for new update out of date dashboard

RobertTStrong commented:
Note: I didn't update the DATA_URL and will do so after a location for the JSON data is determined.

RobertTStrong commented:
Pull request for the notebook
https://github.com/mozilla/mozilla-reports/pull/25

chutten reviewed (approved):
Seems legit to me.

Commit:
Fix Cross-Origin Resource Sharing violation, data url, and notebook location for Bug 1338941 - web page for new update out of date dashboard

RobertTStrong commented:
I managed to add the data to the analysis-output.telemetry.mozilla.org and I changed the url to the proposed location for the mozilla-reports pull request. I forgot that the xhr won't work with cross origin and since I wasn't really happy with how I implemented this I went with what I think is a better solution.

Commit:
Fixed incorrect statement about criteria matching Bug 1338941 - web page for new update out of date dashboard

Commit:
Fixed css to handle change of dynamic naming of class Bug 1338941 - web page for new update out of date dashboard

chutten commented on update-orphaning/index.js:
```diff
@@ -1029,19 +1040,17 @@ function displayLoadErr() {
+  errA.textContent = DATA_URL + document.getElementById("weekly-dropdown").value;
   errDiv.appendChild(errA);
-  errText = document.createTextNode(")");
-  errDiv.appendChild(errText);
```
`errText` is written to but not read?

RobertTStrong replied:
I just removed the parenthesis around the url. Now it is as follows
  var errText = document.createTextNode("URL: ");
  errDiv.appendChild(errText);
  var errA = document.createElement("a");
  errA.setAttribute("href", DATA_URL + document.getElementById("weekly-dropdown").value);
  errA.setAttribute("target", "_blank");
  errA.textContent = DATA_URL + document.getElementById("weekly-dropdown").value;
  errDiv.appendChild(errA);

chutten replied:
Ah, right, crap. I see it now.
